### ISSUE TITLE
Yii::t for JavaScript

### ISSUE THREAD

**Issue by samdark** [yiisoft/yii2#274]
I think translations at clientside are quite a common feature. I think it will be useful if Yii will provide it out of the box.

**Comment by alexandernst** [comment]
Any news on this?

**Comment by samdark** [comment]
It's planned for 2.1.x so we aren't going to work on it for 2.0 unless a pull request provided.

**Comment by alexandernst** [comment]
@samdark I might work on this one. What/how do you think this should work? Here is what I think:

The client side should be very simple, it shouldn't not know anything about string-templating, nor translations, nor anything else. The only thing it should know is what the developer wants to translate and where that text belongs to.

As I see it, there are to possible use-cases. The first one is a string that the developer would like to place somewhere in the DOM. The second one is a string that the developer would like to translate and then use in his own code.

Based on those two assumptions, a possible prototype could look like this:

    yii.t("This is a %s", var_s_test, '#my_span'); 
    yii.t("Another %s with multiple %s", var_s_test, var_s_args, function(err, t){
        console.log(t);
    });

The first line would:

* Place the original string right away in the `#my_span` element in the DOM
* Make an ajax call to the server and wait for a reply
* If there is an error, don't do anything else (or maybe trigger some error event, but either way, keep the original string in the DOM element).
* If there is no error, replace the text from the DOM element with the translated one.

The second line would:

* Make an ajax call to the server and wait  for a reply
* When the reply is ready, call the callback passing the `err` status and the `t` string. If there was an error, `err` would be `true` and the `t` variable would contain the original text. If there was no error, `err` will be `false` and `t` will contain the translated text.

Now, the server part would be even easier. It could consist of a controller with a single method, which would then contain a single line (a call to `Yii::t` with the received arguments).

**Comment by alexandernst** [comment]
@samdark Oh, I forgot one really important thing.

The `yii.t` method should use something like [`memoize`](http://underscorejs.org/#memoize), so further calls containing the exact same string and arguments will not result in another ajax call.

**Comment by samdark** [comment]
I'm not sure any AJAX calls are required. message command could generate json file containing translations and the file could be used then. I was thinking about generating these files from PHP ones and using something like https://github.com/twitter/twitter-cldr-js.

**Comment by Blizzke** [comment]
I added that functionality to one of my bigger projects and you need a couple things more: 
To modify the CLI `MessageController` to also parse .js files (and possible others).

IMO also a category. In the backend this allows you to split translations over various sources. I have a couple number of modules and every module takes care of its own translations.  IMO we should keep things as consistent as possible between front- and backend as well, which means that you should also use the category in the javascript version.

Having a category also allows something else: To return the entire block of strings and their translation at once (perhaps on request only?) so that we don't even need an ajax call per translation. Don't forget that you'll have to work with `Deferred`'s to correctly update whatever you need, so it might save a lot of headaches to fetch the entire translated block at once.

On top of that you probably also want to translate tags that aren't always javascript. I implemented a support for a "data-i18n"-attribute for that purpose (which will replace the html content of whatever tag this is used in. I also have support for a "data-i18n-target", for when you want to replace one of the tags attributes instead of the html content.

Obviously there are a ton of ways to tackle this problem, just wanted to chip in :)

**Comment by alexandernst** [comment]
@samdark @Blizzke So your idea is to actually parse the JS files and search for calls to `yii.t` (or whatever the method's name is)? If so, I'm not really sure how would that work with compressed JS files (see, `var x = yii.t`).

**Comment by Blizzke** [comment]
Imo it should match the existing backend system as closely as possible, so yes.
My yii.t implementation also allows for the same {token} syntax etc. 

As for compressed js, that is (imo) a deployment step and it the message command should run against the regular files. I doubt you'll be writing your original source compressed, no?

**Comment by samdark** [comment]
I don't think one will ever run it against compressed files.

**Comment by alexandernst** [comment]
Ok, makes sense, indeed. Let's see if I can get some spare time and work on this one.

**Comment by Blizzke** [comment]
If needed I can help out or provide some of the things I already added. Let me know

**Comment by alexandernst** [comment]
@Blizzke Yes! Totally! The more, the better! Actually, as I'm currently kind  of short of time you can take the issue if you want to. It sounds like you already got something working :)

**Comment by stepanselyuk** [comment]
My way (which described above, in August 2014) working good in my application.

**Comment by Blizzke** [comment]
I don't mind doing it but perhaps we should iron out first what features people expects of the yii.t function? I doubt my demands are what everyone is looking for.

What do you guys think about:
* MessageController modification to parse *.js files 
* "format" for storing in a json file (on top of the existing ones)
* yii.t functionality that allows you to work with {token}-parts, perhaps some formatting? (do we do categories like the backend or not, imo we should)
* extract html data attributes for html translation (data-i18n, ...)?
* Automatic Loading translations on client-side (assetbundle, ajax?) Or have the user loading via whatever method and registering it with the translation system (yii.translations('category', object))? 
* ...

**Comment by samdark** [comment]
> MessageController modification to parse *.js files

Probably yes. One thing I can say in advance is that it's not reliable to do it via regexp so one should take care of implementing tokenizer for JavaScript like I did for PHP to fix tons of regexp issues.

> "format" for storing in a json file (on top of the existing ones)

Yes.

> yii.t functionality that allows you to work with {token}-parts, perhaps some formatting? (do we do categories like the backend or not, imo we should)

Yes. I'd expect to be able to reuse my PHP messages in JS. So full blown ICU (I've provided a link to twitter's one).

> extract html data attributes for html translation (data-i18n, ...)?

Probably data-t? What are these for? Any examples?

> Automatic Loading translations on client-side (assetbundle, ajax?) Or have the user loading via whatever method and registering it with the translation system (yii.translations('category', object))?

Check how it's done for JUI datapicker currently.

**Comment by nkovacs** [comment]
Take a look at [globalize](https://github.com/jquery/globalize). It has a similar message parsing functionality.

**Comment by samdark** [comment]
Indeed. And it's not adopted by jQuery foundation so it's definitely best possible choice for now.

**Comment by michaelarnauts** [comment]
Is anyone still working on this?

**Comment by AlexGx** [comment]
Have same question

**Comment by Blizzke** [comment]
I actually implemented  most of this for a project, but since there was never any activity on this ticket I haven't bothered with making it releasable. 
I can release the modified `MessageController` (which should work pretty okay as is), but the related javascript is coupled with that project. If I were to release that it would require implementing a couple things yourself. Let me know if there is any interest. I'm not going to continue working on it but I can setup a quick repository for it

**Comment by michaelarnauts** [comment]
@Blizzke if you could release it, I'm interested in the code.

**Comment by samdark** [comment]
Yes, that could be interesting to check. If it will be released as standalone extension, it could be referenced from a guide.

**Comment by Charius** [comment]
And how it'll be compressed via `yii assets` command?

**Comment by Blizzke** [comment]
As I said, I am not going to release this as a fully working extension, I don't have the time nor the enthusiasm for that. I'll setup a repository with all the relevant files and some high level instructions on how to get started with it. Just give me a day or 2 to see if I can scrape together a bit of time.

**Comment by Blizzke** [comment]
If put the "repository" [here](https://github.com/bedezign/yii2-jsi18n)
It provides a controller to parse javascript (and HTML) and deliver the found messages to the regular Message functionality so you can have them written in files / added to the database like regular messages. 

The way I developed it is that it has a jQuery plugin that calls an endpoint in your application (example included) to retrieve the messages as `json`. 
When I developed this I decided I wanted the messages of back- and frontent combined and not just part of the javascript. 

This is offered without any guarantees and probably won't work out of the box. 
The message parsing code and the translation function do work and were tested back then. The part where you fetch them from the backend was just thrown in. The original code worked with a bus that did the processing and did not have any ajax requests.

I will try to answer any issues created for the project but please understand that it has been a while for me :) 
If it is decided you want to turn the repository into an actual yii2 plugin, I will be accepting pull requests, doing releases and adding it to packagist. Otherwise I'd appreciate a shout-out, as requested in the README

**Comment by serhatozles** [comment]
I don't know, what did you do. But I have an idea. We can use as you said "yii.t(category,message)", if we import the all translate data! My idea, we don't need all translate data, developer can use different category for javascript. for example; 

in layout:
```php
Yii::$app->getI18n()->register($this,'javascript',['en-EN','ru-RU']);
Yii::$app->getI18n()->register($this,'alert',['en-EN','tr-TR']);
// maybe there could be have big data for each language, when the language changes by dynamic, we can get the data with ajax? Just an idea.
```
than it sends to javascript like that data:
```javascript
window.yiiTranslateSource = {
    javascript: [
        {
            'bla bla bla': {
                'en-EN': 'bila bila bila',
                'ru-RU': 'bila bila bila2'
            }
        }
    ],
    alert: [
        {
            'Success': {
                'en-EN': 'Success 2',
                'tr-TR': 'Başarılı'
            }
        }
    ]
};
```

and we can use in javascript;
```javascript
yii.t('javascript','Success');
/* we can change language by dynamic */
yii.changeLanguage('en-EN');
```
What do you think?

**Comment by w3lifer** [comment]
My solution: https://github.com/w3lifer/yii2-i18n-js.

**Comment by elzix** [comment]
I was expecting the solution would be to:

- change the file extension to .php
- add the php `<?= Yii::t() ?>` to get translations
- render the file in the main layout file (instead of register as js)

I have this for a CSS file. The code would be printed on the page source, but it would include all the translated messages. However the above solutions seem better since the source would remain without the clatter.

**Comment by elzix** [comment]
Another solution would be to add the translated messages in hidden fields
e.g. `<input type="hidden" id="contact-thanks" value="<?= Yii::t('contact','Thanks!') ?>">`

on the page and have your script use the messages
e.g. `$('#contact .msg-title').html($('#contact-thanks').val());`

**Comment by samdark** [comment]
Since there are multiple working solutions, no plan to handle it in the framework for now.

**Comment by roman-m-grinyov** [comment]
I (@w3lifer) was blocked on GitHub, now I'm on GitLab: https://gitlab.com/w3lifer/yii2-i18n-js.

`composer req w3lifer/yii2-i18n-js` works fine.

**GitLab forever!**
